Testnet/DAOContract/DAOContract/DAOContract.cs

```diff
@@ -0,0 +1,213 @@
+
+    public uint GetNoVotes(uint proposalId) => State.GetUInt32($"NoVotes:{proposalId}");
+    private void SetNoVotes(uint proposalId, uint value) => State.SetUInt32($"NoVotes:{proposalId}", value);
+
+    public uint GetVote(uint proposalId, Address address) => State.GetUInt32($"Vote:{proposalId}:{address}");
+
+    public Proposal GetProposal(uint index) => State.GetStruct<Proposal>($"Proposals:{index}");
+
+    private void SetProposal(uint index, Proposal proposal) => State.SetStruct($"Proposals:{index}", proposal);
+
+    public DAOContract(ISmartContractState state, uint minQuorum)
+        : base(state)
+    {
+        this.MinQuorum = minQuorum;
+        this.Owner = Message.Sender;
+        this.LastProposalId = 1;
+    }
+
+    public uint CreateProposal(Address recipent, ulong amount, uint votingDuration, string description)
+    {
+        var proposal = new Proposal
+        {
+            RequestedAmount = amount,
+            Description = description,
+            Recipient = recipent,
+            Owner = Message.Sender
+        };
+
+        SetProposal(LastProposalId, proposal);
+        SetVotingDeadline(LastProposalId, checked(votingDuration + Block.Number));
+        Log(new ProposalAddedLog
+        {
+            ProposalId = LastProposalId,
+            Recipent = recipent,
+            Amount = amount,
+            Description = description
+        });
+
+        return LastProposalId++;
+    }
+
+    public void Vote(uint proposalId, bool vote)
+    {
+        Assert(IsWhitelisted(Message.Sender), "The caller is not whitelisted.");
+
+        Assert(GetVotingDeadline(proposalId) > Block.Number, "Voting is closed.");
+
+        SetVote(proposalId, Message.Sender, ToVote(vote));
+    }
+
+    private void SetVote(uint proposalId, Address address, Votes vote)
+    {
+        var currentVote = (Votes)GetVote(proposalId, address);
+
+        if (currentVote == vote)
+        {
+            return;
+        }
+
+        Unvote(proposalId, currentVote);
+        State.SetUInt32($"Vote:{proposalId}:{address}", (uint)vote);
+
+        if (vote == Votes.Yes)
+        {
+            SetYesVotes(proposalId, GetYesVotes(proposalId) + 1);
+        }
+        else
+        {
+            SetNoVotes(proposalId, GetNoVotes(proposalId) + 1);
+        }
+    }
+
+    private void Unvote(uint proposalId, Votes currentVote)
+    {
+        switch (currentVote)
+        {
+            case Votes.Yes: SetYesVotes(proposalId, GetYesVotes(proposalId) - 1); break;
+            case Votes.No: SetNoVotes(proposalId, GetNoVotes(proposalId) - 1); break;
+            case Votes.None: break;
+        }
+    }
+
+    private Votes ToVote(bool vote) => vote ? Votes.Yes : Votes.No;
+
+    public void ExecuteProposal(uint proposalId)
+    {
+        var proposal = GetProposal(proposalId);
+
+        Assert(proposal.Owner == Message.Sender, "The proposal can be executed by proposal creator.");
```

Comment:
This might not be necessary, no harm in letting anyone attempt to execute a finished proposal

Reply:
The DAO should be autonomous and not locked by owner to execute an approved, voted on transaction.

Reply:
In here executer is the owner of the proposal, not contract owner so it is autonomous.